### PR TITLE
feat: http client with tracing

### PR DESCRIPTION
**Pull request by adityathebe**
- [x] Vhost
- [x] Tracing
- [x] Retries
- [x] NTLM
- [x] DNS Cache
- [x] We also need to propagate any parent span to upstream calls
- [x] We should also have fine grained options on what to add to trace 
  - [x] request/response headers (sanitized of sensitive headers)
  - [x] request
  - [x] response body (with limits for the content)

### CODE REVIEW

File: context/context.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"k8s.io/client-go/kubernetes"
+)
+
+// TODO:
+type Context interface {
+	Namespace() string
+	Kubernetes() kubernetes.Interface
```
Comment (moshloop) [inline context/context.go:15]: We can't add these as it will then import all the k8s dependencies

File: dns/dns.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+type Cache struct {
+	*marshaler.Marshaler
+}
+
+var cache *Cache
+
+func newCache() (*Cache, error) {
+	bigcacheClient, _ := bigcache.NewBigCache(bigcache.DefaultConfig(60 * time.Minute))
```
Comment (moshloop) [inline dns/dns.go:22]: I think we should just use "github.com/patrickmn/go-cache" directly

File: http/request.go [hunk 3]
```diff
@@ -167,29 +180,74 @@ func (r *Request) createHTTPRequest() (*http.Request, error) {
...
+		delete(r.config.Headers, "Host")
+	}
+
+	if r.config.ConnectTo == "" && r.config.DNSCache {
+		ips, err := dns.CacheLookup(ctx, "A", r.url.Hostname())
+		if len(ips) == 0 {
+			return nil, err
+		}
+
+		r.config.ConnectTo = ips[0].String()
```
Comment (moshloop) [inline http/request.go:214]: should add this as a header

File: http/request.go [hunk 4]
```diff
@@ -167,29 +180,74 @@ func (r *Request) createHTTPRequest() (*http.Request, error) {
 }
 
 // sendRequest sends the request using the given HTTP client
-func (r *Request) sendRequest(client *http.Client, logger logger.Logger, retriesRemaining uint) (*Response, error) {
-	request, err := r.createHTTPRequest()
-	if err != nil {
+func (r *Request) sendRequest(ctx context.Context, client *http.Client, logger logger.Logger, retriesRemaining uint) (*Response, error) {
+	ctx, span := tracer.Start(ctx, r.url.Hostname()) // TODO: Decide on how to name this span
```
Comment (moshloop) [inline http/request.go:184]: I think the tracing should be off by default, only if a tracer and trace name / existing span is specified when creating the client should we trace, 
- We also need to propagate any parent span to upstream calls
- We should also have fine grained options on what to add to trace (req/response headers (sanitized of sensitive headers), request, response body (with limits for the content)

File: http/request.go [hunk 5]
```diff
@@ -167,29 +180,74 @@ func (r *Request) createHTTPRequest() (*http.Request, error) {
...
+	defer span.End()
+
+	span.SetAttributes(attribute.String("method", r.verb))
```
Comment (moshloop) [inline http/request.go:187]: Setting ottributes and errors should be part of the ctx e.g. `ctx.SetAttribute(key, val)` and `ctx.Errorf` e.g. https://github.com/flanksource/incident-commander/commit/58e771879d7464e0c7924dfdd8d68a25160dc726

File: context/context.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+// TODO:
+type Context interface {
+	WithContext(ctx gocontext.Context) Context
+	WithTimeout(timeout time.Duration) (Context, func())
+
+	Errorf(format string, args ...any)
+
+	StartTrace(tracerName string, spanName string) (Context, trace.Span)
```
Comment (moshloop) [inline context/context.go:18]: Can we not have the span as a field on the context? 

and then `ctx.SpanAttribute(key, value)` or `ctx.SpanAttributes(key...)`

Comment (adityathebe) [reply context/context.go:18]: I've gone with a separate approach. About to push.

File: http/request.go (hunk 4 above)
Comment (adityathebe) [reply http/request.go]: @moshloop How would sanitizing sensitive fields from the HTTP header work? Do we ask the user for the list of HTTP headers that needs to be sanitized like this - https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-sanitize-field-names

File: http/middlewares/trace.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+func (t *traceTransport) RoundTripper(rt netHttp.RoundTripper) netHttp.RoundTripper {
+	return http.RoundTripperFunc(func(ogRequest *netHttp.Request) (*netHttp.Response, error) {
+		// According to RoundTripper spec, we shouldn't modify the origin request.
+		req := ogRequest.Clone(ogRequest.Context())
+
+		propagator := propagation.TraceContext{}
+		propagator.Inject(req.Context(), propagation.HeaderCarrier(req.Header))
+
+		_, span := t.tracer.Start(req.Context(), req.URL.Host)
```
Comment (moshloop) [inline http/middlewares/trace.go:106]: The span name should be configurable e.g. upstream-reconcile

File: http/middlewares/trace.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+		defer span.End()
+
+		span.SetAttributes(
+			attribute.String("request.method", req.Method),
+			attribute.String("request.url", req.URL.String()),
+			attribute.String("request.host", req.Host),
+		)
+
+		if t.headers {
```
Comment (moshloop) [inline http/middlewares/trace.go:115]: Request and response headers need to be sanitized of secret values

File: http/examples_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+// TODO: This will be removed or we can keep it as exaples as well
+// Contains some examples
+func TestExample(t *testing.T) {
+	ctx := context.Background()
+
+	client := http.NewClient().
+		BaseURL("https://dummyjson.com").
+		BasicAuth("username", "password").
+		Host("dummyjson.com").
```
Comment (moshloop) [inline http/examples_test.go:24]:
```suggestion
		ConnectTo("dummyjson.com").
```

File: http/examples_test.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		body := &bytes.Buffer{}
+		body.WriteString(`{"title": "test"}`)
+		postReq := client.R(ctx).Header("Scope", "request")
+		response, err := postReq.Post("products/add", body)
+		if err != nil {
+			logger.Fatalf("error: %v", err)
+		}
+
+		var bodyResponse = map[string]any{}
+		if err := response.Into(&bodyResponse); err != nil {
```
Comment (moshloop) [inline http/examples_test.go:39]: Into feels like magic - It should be an explicit From/As JSON

File: http/examples_test.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+	{
+		req := client.R(ctx)
+		response, err := req.Get("products/1")
+		if err != nil {
+			logger.Fatalf("error: %v", err)
+		}
+
+		b, _ := io.ReadAll(response.Body)
+		logger.Infof("GET body: %s %v", string(b), response.IsOK())
```
Comment (moshloop) [inline http/examples_test.go:53]:
```suggestion
		logger.Infof("GET body: %s %v", response.AsString(), response.IsOK())
```

File: http/middlewares/trace.go (hunk 8 above)
Comment (adityathebe) [reply http/middlewares/trace.go:115]: Can you take a look at https://github.com/flanksource/commons/pull/79#discussion_r1349536031?

**Comment by adityathebe** [comment]
@moshloop I've added context to a new PR branched off from this PR. 
https://github.com/flanksource/commons/pull/80

So, far it doesn't seem like it will fit well with our other context. Will add explanations on that PR.

**Pull request merged**